Enhance mgr cache - from cache based on ttl to cache based on changes

NitzanMordhai opened the pull request:
### 1. Introduction & Problem Statement
The Ceph Manager (Mgr) provides a caching mechanism to improve the performance of Python modules that frequently request cluster data like the OSDMap. Currently, this cache relies on a fixed Time-To-Live (TTL), controlled by the mgr_ttl_cache_expire_seconds setting.
While this approach reduces load, it presents a fundamental trade-off: a long TTL risks modules operating on stale data, while a short TTL causes unnecessary resource consumption by repeatedly regenerating data that has not changed. This document proposes an enhancement to replace the TTL-based expiration with a more intelligent, event-driven invalidation strategy.

### 2. Background: The Current Caching Mechanism
To understand the cache, it's essential to trace the full API call flow from a Python module to the C++ backend and back.
Flow:

1. Python Module Call: A mgr module requests cluster data, for example: osd_map = mgr.get('osd_map').
2. Mgr API Wrapper: This call is handled by the get() method in the MgrModule class in mgr_module.py. This method is exposed to all modules via the @API.expose decorator.
3. Internal C-API Bridge: The Python get() method calls self._ceph_get('osd_map'). This internal function is defined in C++ (BaseMgrModule.cc) and mapped via a PyMethodDef table.
4. C++ Dispatch: The bridge function (ceph_state_get) parses the string argument ('osd_map') and passes it to the C++ method cacheable_get_python().
5. Caching and Data Retrieval: This is the core caching layer. cacheable_get_python attempts to retrieve the data from the TTLCache. This results in either a "cache hit" or a "cache miss."
6. Return to Python: The resulting PyObject* (a pointer to a Python object) is returned back up the call stack to the mgr_module.py wrapper.
7. Final Processing: The Python wrapper checks the returned object. If it's a bytes object (which occurs on a cache hit or a miss where caching is enabled), it uses json.loads() to deserialize it into a standard Python dictionary.
Final Return: The ready-to-use Python dictionary is returned to the original module code.

 **Cache Logic**
The TTLCache class that underpins this mechanism is specifically designed to handle Python objects safely in a C++ environment. It uses Py_INCREF when an object is inserted or retrieved and Py_DECREF when an object is erased, correctly managing the object's lifecycle and preventing memory leaks.
The caching logic operates in two main scenarios:

- Cache Miss: If an item is not in the cache, get_python() is called. It uses PyJSONFormatter to serialize the map data into a Python bytes object. This PyObject* is then inserted into the TTLCache.

- Cache Hit: If a valid item is in the cache, the TTLCache retrieves the existing PyObject* and returns it, skipping the expensive data generation step.

### 
8. Limitations of the Current TTL-Based Approach
The reliance on a fixed TTL, while functional, leads to two significant problems:

- Stale Data: If the TTL is set to 10 seconds, a module may continue to see an old OSDMap for up to 10 seconds after a new map has been received by the manager, leading to potentially incorrect behavior.

- Wasted Resources: If the TTL is set to a low value, the manager will frequently expire and rebuild cached objects (like pg_stats or osd_map) even if the underlying data has not changed, consuming unnecessary CPU cycles on serialization.


### 4. Proposed Enhancement: Event-Driven Invalidation
This proposal moves from a passive time-based expiration model to an active, event-driven one, fundamentally improving how we handle cached data. This involves several key changes:

1. Rename TTLCache to MgrCache: Since we are removing the Time-To-Live logic, the name TTLCache is no longer accurate. We will rename it to MgrCache to better reflect its new role as a persistent, event-driven cache for the manager.
2. Remove ceph options that are no longer needed for TTL cache.
3. Make Caching the Default Path: With an intelligent invalidation strategy, there is no longer a need for a non-cached path. We can enable caching by default and remove the PyJSONFormatter and its associated logic. This simplifies the codebase by making the PyFormatter path the single, standard way to generate data objects. With that change, we no longer need to load json format and build new python objects, we are saving python objects in the cache.
4. Implement Event-Driven Invalidation Logic:

- The manager daemon already receives notifications from the Ceph Monitors whenever a cluster map is updated. The existing ActivePyModules::notify_all method will serve as the entry point for these updates.

- Inside notify_all, we will call a new method, such as MgrCache::invalidate(map_name), when a relevant map update is received.

- This invalidate method will immediately erase the specific entry (e.g., 'osd_map') from the MgrCache.

- The next time a module calls mgr.get('osd_map'), the request will result in a cache miss. This will force the API to generate a new, up-to-date object from the fresh map and repopulate the MgrCache.

5. Implement a new Ceph CLI command, such as: 

ceph mgr cache flush [map_name]

To address cache issues, a targeted flush of specific maps will be implemented, eliminating the need for a full Manager restart.

### 5. Benefits
This enhancement provides significant advantages over the current system:

- Guaranteed Data Freshness: Modules will have access to cluster data that is current as of the last received map, eliminating the stale data problem.

- Increased Efficiency: Data that has not changed will remain in the cache indefinitely, preventing wasteful reprocessing and reducing CPU load on the manager node.

- Simplified Configuration: This change removes the need for administrators to find a suitable balance for the mgr_ttl_cache_expire_seconds configuration value.

Tracker: https://tracker.ceph.com/issues/72447

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

NitzanMordhai committed:
mgr: replace TTLCache with MgrMapCache and protect api with readonly

This patch removes the old TTLCache implementation and introduces
a new generic MgrMapCache driven by a runtime toggle:

- Add `mgr_map_cache_enabled` config option in global.yaml
- Swap out `ttl_cache` for `api_cache` (MgrMapCache) in ActivePyModules
- Update cacheable_get_python() and get_python() to use LFU‐based api_cache
- add new get_mutable parameter to the get api call to get a copy.
- Invalidate api_cache on notify_all events
- Remove all TTLCache headers, sources, and tests
- Include MgrMapCache.cc in CMakeLists and update BaseMgrModule bindings
- Improve logging around cache hits, misses, and state changes

- ActivePyModules
  * Remove unused update_cache_metrics()
  * Log cache hits/misses inline and only insert into cache when
    enabled+cacheable (with proper Py_INCREF)
  * Switch get_python() to use PyFormatterRO for cacheable routes, PyFormatter otherwise

- MgrMapCache/LFUCache
  * Add can_read_cache()/can_write_cache() helpers and use const& for key parameters
  * Guard perf counter increments and improve debug logging

- PyFormatter
  * Add PyFormatterRO subclass that freezes dicts/lists into read-only
    proxies on the fly

- Python mgr_module
  * Simplify get() to return raw result

This change ensures immutable JSON output on cache hits and tightens up cache logic.

mgr/cli: add cache flush command with proper status reporting

Allow operators to invalidate individual mgr Python caches at runtime
without restarting the manager. Introduces a new CLI command:

  $ ceph mgr cli cache flush <map-name>

which returns success or a clear error if the named cache entry doesn’t
exist or isn’t cacheable. This makes it easy to drop stale cached maps
(e.g. osd_map, mon_map) on demand.

Fixes: https://tracker.ceph.com/issues/72447
Signed-off-by: Nitzan Mordechai <nmordec@ibm.com>

mgr: add new unit tests for MgrMapCache

- Guard against null perf‐counter before calling inc(), preventing crashes
- Add “foo” to allowed_keys list (for test coverage)
- Rename and refocus the CMake test target from TTLCache to MgrMapCache
- Introduce test_mgrmapcache.cc with LFUCache tests.
- Remove the obsolete test_ttlcache.cc

Fixes: https://tracker.ceph.com/issues/72447
Signed-off-by: Nitzan Mordechai <nmordec@ibm.com>

mgr/test_cache: add new tests

adding new unit-test for mgrcache

Fixes: https://tracker.ceph.com/issues/72447
Signed-off-by: Nitzan Mordechai <nmordec@ibm.com>

NitzanMordhai committed:
mgr/insights: fix read-only copy before change

since we are modifying read-only python object, we need to copy it first

Fixes: https://tracker.ceph.com/issues/72447
Signed-off-by: Nitzan Mordechai <nmordec@ibm.com>

NitzanMordhai committed:
mgr/telemetry: get mutable copy for pool stats

Since we are changing the 'application' for the report,
we need non-RO, in case of cached api call.
using 'pool_stats' map directly to avoid copy of the pg_dump
that can be huge.

Fixes: https://tracker.ceph.com/issues/72447
Signed-off-by: Nitzan Mordechai <nmordec@ibm.com>

NitzanMordhai committed:
mgr/cli_api: pretty_json for mappingproxy fix

since we are modifying read-only python object, we need to copy it first

Fixes: https://tracker.ceph.com/issues/72447
Signed-off-by: Nitzan Mordechai <nmordec@ibm.com>

sseshasa commented on src/mgr/MgrMapCache.h:
```diff
@@ -23,8 +23,16 @@ class LFUCache {
   mutable std::shared_mutex m;
 
-  void mark_miss() { misses++; perfcounter->inc(l_mgr_cache_miss); }
-  void mark_hit() { hits++; perfcounter->inc(l_mgr_cache_hit); }
+  void mark_miss() {
+    misses++;
+    if (perfcounter)
+      perfcounter->inc(l_mgr_cache_miss);
+  }
+  void mark_hit() {
+    hits++;
+    if (perfcounter)
+      perfcounter->inc(l_mgr_cache_hit);
+  }
```
This change probably belongs to the previous commit that introduced MgrMapCache.h.

sseshasa commented on src/test/mgr/test_mgrmapcache.cc:
```diff
@@ -0,0 +1,108 @@
+
```
Please add the Copyright header.

sseshasa commented on src/mgr/ActivePyModules.cc:
```diff
@@ -185,41 +185,55 @@ PyObject *ActivePyModules::get_daemon_status_python(
+int ActivePyModules::ceph_cache_map_erase(const std::string &what)
+{
+  if (!api_cache.exists(what)) {
+    dout(10) << __func__ << " what: " << what << " not in cache" << dendl;
```
`dout_prefix` already includes `__func__`, and therefore may be omitted here and in other places.

sseshasa commented on src/mgr/ActivePyModules.cc:
```diff
@@ -185,41 +185,55 @@ PyObject *ActivePyModules::get_daemon_status_python(
-    auto hit_miss_ratio = ttl_cache.get_hit_miss_ratio();
-    perfcounter->set(l_mgr_cache_hit, hit_miss_ratio.first);
-    perfcounter->set(l_mgr_cache_miss, hit_miss_ratio.second);
+int ActivePyModules::ceph_cache_map_erase(const std::string &what)
```
Can `std::string_view` be used here instead of `const std::string &`?

NitzanMordhai replied:
yes, we can do it, ceph_cache_map_erase only reads the string, not using it for anything else.

NitzanMordhai replied:
actually i can change few more calls as well

sseshasa replied:
Looking at the modified functions alone, `ActivePyModules::get_python()` is another one that can be changed to use `std::string_view`.

sseshasa commented on src/mgr/MgrMapCache.h:
```diff
@@ -0,0 +1,104 @@
+#pragma once
```
Please include Copyright header here and in any new file introduced.

sseshasa commented on src/mgr/MgrMapCache.h:
```diff
@@ -0,0 +1,104 @@
+
+template<class Key, class Value>
+class LFUCache {
+  struct Entry { Value val; size_t hits = 0;};
```
Suggest restructuring the struct declaration to conventional format:

```
 struct Entry {
   Value val;
   size_t hits = 0;
 };
```

sseshasa commented on src/mgr/MgrMapCache.h:
```diff
@@ -0,0 +1,104 @@
+  
+  std::unordered_set<std::string> allowed_keys = {"osd_map", "pg_dump", "pg_stats",
+    "mon_status", "mgr_map", "devices", "osd_metadata",
+    "mds_metadata", "config", "foo"};
```
Perhaps a comment to convey that "foo" is included for unit tests would be good.

sseshasa reviewed (changes requested):
Added a couple of comments during a cursory first pass. Will continue reviewing.

sseshasa commented on src/mgr/MgrMapCache.h:
```diff
@@ -0,0 +1,104 @@
+  mutable std::shared_mutex m;
+
+  void mark_miss() { misses++; perfcounter->inc(l_mgr_cache_miss); }
+  void mark_hit() { hits++; perfcounter->inc(l_mgr_cache_hit); }
```
As with the struct above, restructure these functions and others below according to the Ceph conventional format for easier reading.

sseshasa commented on src/mgr/MgrMapCache.h:
```diff
@@ -0,0 +1,104 @@
+  void set_enabled(bool e) { enabled.store(e); if (!e) { clear(); } }
+  bool is_enabled() const noexcept { return enabled.load(); }
+  int size() { return cache_data.size(); }
+  bool is_cacheable(Key key) const noexcept { return allowed_keys.count(key) > 0; }
```
Couple of comments with `is_cacheable()`:

1.  As Key is a string, `const Key& key` can be passed to avoid making a copy.
2. Since `allowed_keys` is of type `std::unordered_set`, I think `find()` would be faster than `count()` it will stop searching at the first occurrence since the set has unique entries, . Therefore, the check can be changed to: `allowed_keys.find(key) != allowed_keys.end();`
3. To protect against unintended consequences if `Key` is not a string type, would a check be better to ensure `Key` is of type `std::string`? A static assert using a combination of `std::is_same_v` and `std::decay_t` would help here I think.

sseshasa replied:
Thinking a bit further, since Key template parameter will always be `std::string`, it's probably meaningless to treat it as a template. The class is restricting Key to be a string and so defeats the purpose of a generic template. The Key parameter can be removed entirely and change the unordered_map to simply be:

`std::unordered_map<std::string, Entry> cache_data;`

sseshasa commented on src/mgr/MgrMapCache.h:
```diff
@@ -0,0 +1,104 @@
+  explicit LFUCache(size_t cap = UINT16_MAX, const bool ena = true) : capacity{cap}, enabled{ena} {}
+  void set_enabled(bool e) { enabled.store(e); if (!e) { clear(); } }
+  bool is_enabled() const noexcept { return enabled.load(); }
+  int size() { return cache_data.size(); }
```
`size()` must hold the  `std::shared_lock`.
Also the method can be made `const` and can return `size_t` to be consistent with `capacity` type.

sseshasa commented on src/mgr/MgrMapCache.h:
```diff
@@ -0,0 +1,104 @@
+
+  Value get(const Key& k);
+  uint64_t get_hits() const { return hits.load(); }
+  uint64_t get_misses() const { return misses.load(); }
```
Shouldn't these functions also be `const noexcept` for const correctness?

NitzanMordhai replied:
yes, they are. fixing

sseshasa commented on src/mgr/MgrMapCache.h:
```diff
@@ -0,0 +1,104 @@
+  struct Entry { Value val; size_t hits = 0;};
+  std::atomic<uint64_t> hits{0}, misses{0};
+  
+  std::unordered_set<std::string> allowed_keys = {"osd_map", "pg_dump", "pg_stats",
```
Consider making this set `const` if this is going to be a static list and won't change during runtime. 

Consider making LFUCache more generic by passing the allowed keys to the constructor instead of hardcoding the keys.

NitzanMordhai replied:
I don't think we should pass it to the contractor, it doesn't change that frequently that it needs to pass as param

sseshasa replied:
I meant passing the keys as a list to the constructor. This way the class can be made more generic and it can be re-used to work with any other list of string keys and not restricted to the hard coded values.

sseshasa commented on src/mgr/MgrMapCache.h:
```diff
@@ -0,0 +1,104 @@
+  bool can_read_cache(const Key &key) const noexcept { return is_enabled() && is_cacheable(key) && exists(key); }
+  bool can_write_cache(const Key &key) const noexcept { return is_enabled() && is_cacheable(key); }
+
+  bool try_get(const Key &k, Value* out, bool count_hit = true);
```
Should `try_get()` also be marked as `noexcept` considering the child class marks is as `noexcept`? Otherwise, it must be ensured that the base class does not throw.

sseshasa reviewed (changes requested):
Added a few comments on MgrMapCache.h. Still reviewing the rest of the PR.

sseshasa commented on src/test/mgr/test_ttlcache.cc:
```diff
@@ -1,5 +1,6 @@
 
+#include "mgr/MgrMapCache.h"
 #include "gtest/gtest.h"
 #include "mgr/TTLCache.h"
```
I think this file should be removed.

NitzanMordhai replied:
i don't know why I even keep that test file. We no longer have ttlcache. removing it completely

sseshasa commented on src/mgr/MgrMapCache.h:
```diff
@@ -0,0 +1,104 @@
+public:
+  explicit MgrMapCache(uint16_t sz=UINT16_MAX);
+  ~MgrMapCache() { this->clear(); }
+  bool try_get(const Key& k, Value* out, bool count_hit=true) noexcept { return CacheImp::try_get(k,out,count_hit); }
```
nit: Add space between parameters in `try_get()`.

sseshasa commented on src/mgr/MgrMapCache.h:
```diff
@@ -0,0 +1,104 @@
+  void mark_hit() { hits++; perfcounter->inc(l_mgr_cache_hit); }
+
+public:
+  explicit LFUCache(size_t cap = UINT16_MAX, const bool ena = true) : capacity{cap}, enabled{ena} {}
```
To avoid undefined behaviour, a virtual destructor can be added to the `LFUCache`.

sseshasa commented on src/mgr/MgrMapCache.h:
```diff
@@ -0,0 +1,104 @@
+  using CacheImp = LFUCache<Key, Value>;
+public:
+  explicit MgrMapCache(uint16_t sz=UINT16_MAX);
+  ~MgrMapCache() { this->clear(); }
```
The observer needs to be removed as part of the destructors here and in the other specialization of this class.

```
g_conf().remove_observer(this); 
clear();
```

sseshasa commented on src/mgr/MgrMapCache.cc:
```diff
@@ -0,0 +1,163 @@
+
+  // at capacity? evict the lowest‐hit key
+  if (cache_data.size() >= capacity) {
+    auto min_it = std::min_element(cache_data.begin(), cache_data.end(),
```
Here `std::min_element()` is used to iterate over the entire unordered_map every time cache hits capacity in which case this is a O(N) operation and cause latency spikes. You'd probably need a mechanism to maintain keys based on the hit count. The keys with the lowest hit count can then be looked up in O(1) time to evict.

sseshasa replied:
This can probably be ignored for now as the number of keys to scan is very small. But please correct me if this is something to consider at a later point.

sseshasa commented on src/mgr/MgrMapCache.cc:
```diff
@@ -0,0 +1,163 @@
+void MgrMapCache<Key, PyObject*>::insert(const Key &key, PyObject* value) {
+  if (!this->can_write_cache(key)) { return; }
+  ceph_assert(PyGILState_Check());
+  CacheImp::insert(std::move(key), value);
```
If `CacheImp::insert()` results in an eviction, then shouldn't the `Py_DECREF(v)` be invoked for the removed object? In the present code, it looks like there will be a ref leak.  `CacheImp::insert()` should probably return the removed object (if evict was necessary or if the object exists and was overwritten) and then `Py_DECREF(v)` should be invoked on the object.

For e.g., 
```
...
ceph_assert(PyGILState_Check());
Py_INCREF(value); 
PyObject* evicted_obj = CacheImp::insert(key, value);
if (evicted_obj) {
  Py_DECREF(evicted_obj);
}
...
```

My earlier  comment on `LFUCache::insert()` ties into this and should maintain the keys based on hit count.

sseshasa replied:
Since this is new area for me, I was looking closer and reading up on Python and C++ interactions on how python objects are managed.  

Please correct me if I am wrong. Even with the current modified changes, there could be a subtle bug related to memory leaks and races exposed when two threads perform insert operation when the capacity is full. 

Assume thread A performs update of an existing key and thread B adds a new key.

Just after  thread A 'schedules' the `Py_DECREF` via `Py_AddPendingCall()` and before it executes with the cache still at full capacity, thread B adds a completely new key to the cache. Since this will be cache miss, it finds a key with the lowest hit count and erase it from the map and returns the evicted value. Note that the object identified to be overwritten by thread A still has refcount = 1. In this scenario a couple of cases can happen, one of which can obviously be:

[thread B evicts the same keys of thread A]: If eviction logic accidentally targets the newly inserted Object or targets original Object of thread A, thread B hands a pointer to the wrapper. Assuming thread B targets the original object of thread A, then Python's memory manager will deallocate it. Eventually when thread A's pending call wakes up and calls Py_DECREF() on that exact same pointer address, a crash can ensue.

To resolve this, the base class (LFUCache) `insert()` can return a `std::pair` of optionals instead of one. The first key can be set to the object being overwritten (cache hit). The second optional can be set to the evicted item (cache miss due to full capacity). This ensures that the race is handled.

For e.g.:
`std::pair<std::optional<Value>, std::optional<Value>> insert(std::string_view key, Value value);`

Please let me know if my understanding is correct or if I am totally wrong here.

sseshasa commented on src/mgr/MgrMapCache.cc:
```diff
@@ -0,0 +1,163 @@
+}
+
+template<class Key, class Value>
+Value LFUCache<Key,Value>::get(const Key& k) {
```
Both `LFUCache<Key,Value>::get()` and `LFUCache<Key,Value>::try_get()` use the `unique_lock` which is correct since they modify `Entry::hits`. But isn't the read concurrency affected even though a shared_mutex is used? Perhaps making `Entry::hits`  `std::atomic<size_t>` will help allow concurrent reads.

NitzanMordhai replied:
std::atomic will make the Entry much more complex; let me rethink that change and consider a different solution for it.
hits could be approximate and not need to be exact, it just for messaging - i'll do something about it

sseshasa reviewed (changes requested):
Added a few more comments mainly around the MgrMapCache implementation. I will take a look again once the existing comments are fixed or clarified.

NitzanMordhai commented:
@sseshasa I made the changes, please check

sseshasa commented on src/mgr/ActivePyModules.cc:
```diff
@@ -185,41 +185,55 @@ PyObject *ActivePyModules::get_daemon_status_python(
 }
 
-PyObject *ActivePyModules::get_python(const std::string &what)
+PyObject *ActivePyModules::get_python(const std::string &what, const bool get_mutable)
```
This is another method that can be changed to use `std::string_view` .

sseshasa commented on src/mgr/ActivePyModules.cc:
```diff
@@ -523,10 +537,11 @@ PyObject *ActivePyModules::get_python(const std::string &what)
+    return py_formatter_ro.get();
   } else {
-    return pf.get();
+    return py_formatter.get();
```
Since a reference to the Formatter is already set above based on `use_cache`, `f.get()` should work here.

sseshasa commented on src/mgr/BaseMgrModule.cc:
```diff
@@ -384,11 +384,31 @@ static PyObject*
+  int get_mutable = 0; // Default to False
+  dout(10) << __func__ << " called" << dendl;
+  if (!PyArg_ParseTuple(args, "s|i:ceph_state_get", &what, &get_mutable)) {
+    dout(0) << __func__ << " Invalid args!" << dendl;
```
Can use `derr` to be consistent with the rest of the file.

sseshasa commented on src/mgr/BaseMgrModule.cc:
```diff
@@ -384,11 +384,31 @@ static PyObject*
+  char *what = NULL;
+  dout(10) << __func__ << " called" << dendl;
+  if (!PyArg_ParseTuple(args, "s:ceph_cache_map_erase", &what)) {
+    dout(0) << __func__ << " Invalid args!" << dendl;
```
derr?

sseshasa commented on src/mgr/BaseMgrModule.cc:
```diff
@@ -1588,7 +1608,10 @@ PyMethodDef BaseMgrModule_methods[] = {
+   {"_ceph_erase", (PyCFunction)ceph_cache_map_erase, METH_VARARGS,
+   "Erase a cached python map"},
+
+   {"_ceph_notify_all", (PyCFunction)ceph_notify_all, METH_VARARGS,
```
Nit: Alignment is off by a space or two.

sseshasa commented on src/mgr/MgrMapCache.cc:
```diff
@@ -0,0 +1,223 @@
+template <class Key>
+MgrMapCache<Key, PyObject*>::~MgrMapCache() {
+  g_conf().remove_observer(this);
+  CacheImp::clear();
```
Shoudn't this just call the subclass `clear()` instead of the base class `CacheImp::clear()`? Calling base class `clear()` drops the `PyObject*` map entries without protection causing issues down the line.

sseshasa commented on src/mgr/MgrMapCache.cc:
```diff
@@ -0,0 +1,223 @@
+template<class Key, class Value>
+bool LFUCache<Key,Value>::extract(std::string_view k, Value* out) noexcept {
+  std::unique_lock<std::shared_mutex> l(m);
+  auto it = cache_data.find(std::string(k));
```
Passing `std::string_view` and then converting to `std::string` causes allocations on heap which is expensive considering the number of frequent calls to `insert()`, `is_cacheable()` and `can_read_cache()`.

To address this, a `std::vector<std::string_view> allowed_keys` can be defined instead and sorted once in the LFUCache constructor which ensures O(log N) look-ups. WDYT?

This needs to be changed in each function that creates the string.

NitzanMordhai replied:
Good point. 
But instead of a sorted vector, I went with C++20 heterogeneous lookup - added a StringViewHash with is_transparent and plugged it into both allowed_keys and cache_data. Now all the find() calls accept std::string_view directly, no std::string construction on lookups. The only allocation left is in insert() when storing a genuinely new key, which is unavoidable.

sseshasa commented on src/mgr/MgrMapCache.cc:
```diff
@@ -0,0 +1,223 @@
+template<class Key, class Value>
+bool LFUCache<Key,Value>::try_get(std::string_view k, Value* out, bool count_hit) noexcept {
+  std::unique_lock l(m);
+  auto it = cache_data.find(std::string(k));
```
Use `auto it = cache_data.find(k);`

NitzanMordhai replied:
find() can't find a string_view, but i want to use another option, see the above comment

sseshasa commented on src/mgr/MgrMapCache.cc:
```diff
@@ -0,0 +1,223 @@
+  auto it = cache_data.find(key_str);
+  if (it != cache_data.end()) {
+    // If key exists, return old value for cleanup
+    Value old_val = it->second.val;
```
`Value old_val = std::move(it->second.val);` Based on my comments about the race with `insert()`, the LHS would be one of the `std::optional`.

NitzanMordhai replied:
yes, i'll use std::optional

sseshasa commented on src/mgr/MgrMapCache.cc:
```diff
@@ -0,0 +1,223 @@
+        return a.second.hits.load(std::memory_order_relaxed) <
+               b.second.hits.load(std::memory_order_relaxed);
+    });
+    evicted = min_it->second.val;
```
`evicted = std::move(min_it->second.val);`. Based on my comments about the race with `insert()`, the LHS would be one of the `std::optional`.

sseshasa commented on src/mgr/MgrMapCache.cc:
```diff
@@ -0,0 +1,223 @@
+    PyObject* old_obj = evicted.value();
+    if (Py_AddPendingCall(+[](void* p){ Py_DECREF((PyObject*)p); return 0; }, old_obj) != 0) {
+      Py_DECREF(old_obj);
+    }
```
Should the callback function acquire the GIL to be safer like below?
```
if (Py_AddPendingCall(+[](void* p){ Py_DECREF((PyObject*)p); return 0; }, old_obj) != 0) {
      PyGILState_STATE st = PyGILState_Ensure();
      Py_DECREF(old_obj);
      PyGILState_Release(st);
}
```

sseshasa commented on src/mgr/MgrMapCache.cc:
```diff
@@ -0,0 +1,223 @@
+
+template<class Key, class Value>
+bool LFUCache<Key,Value>::try_get(std::string_view k, Value* out, bool count_hit) noexcept {
+  std::unique_lock l(m);
```
There's still a concurrent reading problem here by using unique_lock. This should be `shared_lock`.

sseshasa commented on src/mgr/MgrMapCache.cc:
```diff
@@ -0,0 +1,223 @@
+  if (count_hit) {
+    it->second.hits++;
+    mark_hit();
+  }
```
Since we changed to use `shared_lock`, this section should probably be:
```
if (it == cache_data.end()) {
  if (count_hit) {
    mark_miss(); 
  }
  return false;
}
if (count_hit) {
  it->second.hits.fetch_add(1, std::memory_order_relaxed); // Atomic, allocation-free, safe increment
  mark_hit();
}
```

sseshasa commented on src/mgr/MgrMapCache.h:
```diff
@@ -0,0 +1,173 @@
+#include "mgr/mgr_perf_counters.h"
+#include "PyUtil.h"
+
+template<class Key, class Value>
```
If you consider removing the Key template parameter as per one of my comments, you can also consider moving most of the functionality to the base LFUCache class. The generic `MgrMapCache<Value>` can act as a lightweight passthrough abstraction and free of Python logic. This layer can just handle the `handle_conf_changes()`,  `erase(std::string_view k)` which can just invoke `(void)CacheImp::extract(k, &v);` and the destructor of course which should call `remove_observer` and `this->clear() `as usual. The constructor should be similar to the constructor of the PyObject* specialization.

All the complex GIL related operations can be delegated to the `MgrMapCache<PyObject*>` specialization of the `LFUCache`.

sseshasa commented on src/mgr/MgrMapCache.h:
```diff
@@ -0,0 +1,173 @@
+    Value val;
+    mutable std::atomic<size_t> hits;
+    
+    Entry() : hits(0) {}
```
change to `Entry() = default;`

sseshasa commented on src/mgr/MgrMapCache.h:
```diff
@@ -0,0 +1,173 @@
+
+template<class Key, class Value>
+class LFUCache {
+  static_assert(std::is_same_v<std::decay_t<Key>, std::string>, 
```
This static_assert can go away if the `Key` template parameter is removed.

sseshasa reviewed (commented):
Some more comments that need looking into. PTAL.

sseshasa commented on src/mgr/MgrMapCache.h:
```diff
@@ -0,0 +1,173 @@
+
+  struct Entry {
+    Value val;
+    mutable std::atomic<size_t> hits;
```
`mutable std::atomic<size_t> hits{0};`

NitzanMordhai commented:
Good catch on thinking through the race. 
The unique_lock in insert() actually prevents the scenario you described - 
 once Thread A moves ptr_old out of the map, the lock is still held, 
 so Thread B can't enter the eviction path until Thread A is done. 
 By that point ptr_old is no longer in cache_data, so Thread B's eviction scan can never return the same pointer. 

The pair return we implemented is still the right design for clarity, but the mutual exclusion is what keeps the refcount accounting safe.

NitzanMordhai commented:
@sseshasa Thanks for the review, i made the changes. Please take a look

sseshasa commented on src/mgr/ActivePyModules.cc:
```diff
@@ -185,41 +185,55 @@ PyObject *ActivePyModules::get_daemon_status_python(
+    !get_mutable &&
+    api_cache.is_enabled() &&
+    api_cache.is_cacheable(what) &&
+    PyGILState_Check();
```
Should this be wrapped in a `ceph_assert()`?

NitzanMordhai replied:
Not needed here  PyGILState_Check() is used as a condition rather than an assertion by design.
get_python() can be called without the GIL, in which case we just skip the cache and fall through to the normal path. An assert would crash the daemon in valid call scenarios.

sseshasa commented on src/mgr/MgrMapCache.cc:
```diff
@@ -0,0 +1,233 @@
+    return {std::nullopt, std::nullopt};
+  }
+
+  std::unique_lock<std::shared_mutex> l(m);
```
Couple of comments here:

1. We probably need a check to see if the cache state has changed while waiting. Specifically, the atomic `enabled` flag can be checked here. If state flipped while waiting, `std::nullopt` pair must be returned. Therefore, with the above check we are ensuring that `{std::nullopt, std::nullopt}` is returned only in the following cases:
 - Successful rejection (cannot write to cache or cache is disabled).
 - Successful insertion (no overwrite or eviction).

In other cases, we either return the old value or the evicted item.

2. Therefore, there are 3 possible outcomes with the Base insert operation. The specialization needs to handle the 3 cases. The current result returned leaves scope for ambiguity and could result in an invalid `Py_DECREF()` due to a race (as explained in the comment on the specialization ` MgrMapCache<PyObject*>::insert()`  implemenation.

The Base insert probably needs an additional `bool` in the response to indicate if an item was inserted or not. Based on this, the specialization can take a clear decision on the next step. My suggestion is to either return a 

`std::tuple<bool, std::optional<value>, std::optional<value>>` 

where bool indicates if the value was successfully inserted or not.

or better still a struct for e.g: 

```
struct InsertRes {
    bool inserted{false};           // True if the value was saved to the cache
    std::optional<Value> replaced;  //if an old value was overwritten
    std::optional<Value> evicted;   // if an old value was evicted
  };
```

NitzanMordhai replied:
yes, sounds like good changes to me, let me add them and test it

NitzanMordhai replied:
Also adding some unit-test for it as well

sseshasa replied:
LGTM.

sseshasa commented on src/mgr/MgrMapCache.cc:
```diff
@@ -0,0 +1,233 @@
+  // If the base class rejected the insert (e.g. cache disabled between our
+  // can_write_cache check and the lock acquisition inside insert()), undo the
+  // INCREF to avoid a reference leak.
+  if (!overwritten && !evicted && !this->exists(key)) {
```
1. Probably safer to use `!overwritten.has_value() && !evicted.has_value()`.

2. Also, there's still a scope for a race here. 
 - Assume thread A successfully inserted a new item and returns `{std::nullopt, std::nullopt}`.
 - Thread A moves to the check involving ` !this->exists(key)`. 
 - Thread B preempts thread A just before it can evaluate `!this->exists(key)`.
 - Thread B calls `clear()` or `erase(key)`. It acquires the `unique_lock`, wipes the map and releases the lock.
 - Thread A resumes execution and evaluates  ` !this->exists(key)` which is `True` in this case and executes `Py_DECREF(value)`.
 - The above will lead to memory related issues down the line.

To remove the ambiguity of the Base `insert()`, its return value can include a a bool as suggested above to clearly decide the next step.

NitzanMordhai replied:
1. now that using InsertRes i don't need that check we already setting InsertRes.inserted
2. again, we are using InsertRes, which removes that race concern

sseshasa commented on src/mgr/MgrMapCache.cc:
```diff
@@ -0,0 +1,233 @@
+template<class Value>
+MgrMapCache<Value>::~MgrMapCache() {
+  g_conf().remove_observer(this);
+  this->clear();
```
`this->clear();` is unnecessary here as the `PyObject` specialization calls its own `clear()` which performs `drain()` and decrements all Python references under GIL.

Having the base call `this->clear()` here exposes a race between the time the specialized `clear()` and the base `clear()` is called. For e.g., if a new entry is added between the two calls, the base `clear()` can wipe out the new data.

NitzanMordhai replied:
the generic template destructor doesn't participate in the PyObject* destruction chain at all  full specialization is completely independent. 
The this->clear() in the generic template is redundant, yes, but there's no double-clear race. removing the clear() call

sseshasa commented on src/mgr/MgrMapCache.cc:
```diff
@@ -0,0 +1,233 @@
+}
+
+
+template class MgrMapCache<PyObject*>;
```
This is probably not needed as it is already specilized.

sseshasa reviewed (commented):
Looks much better. Added a few more comments around the MgrMapCache.h/cc. Getting close!

sseshasa commented on src/mgr/MgrMapCache.cc:
```diff
@@ -0,0 +1,232 @@
+
+  std::unique_lock<std::shared_mutex> l(m);
+  // Re-check enabled after acquiring lock: state may have flipped while waiting.
+  if (!enabled.load()) {
```
Probably need to add `std::memory_order_relaxed` optimization here.

sseshasa commented on src/mgr/MgrMapCache.cc:
```diff
@@ -0,0 +1,232 @@
+  mark_miss();
+
+  InsertRes res{true};
+  if (cache_data.size() >= capacity) {
```
To avoid `std::min_element` working on an empty container and causing seg faults when trying to access `min_it->second.val`, the 'if' condition can be changed to:

`if (cache_data.size() >= capacity && !cache_data.empty())`

NitzanMordhai replied:
Added, just a note: for any capacity >= 1, the original condition was safe - size >= 1 implies non-empty. The only vulnerable case is capacity == 0, where 0 >= 0 is true on an empty map. The guard !cache_data.empty() closes that gap cleanly.

sseshasa commented on src/mgr/MgrMapCache.cc:
```diff
@@ -0,0 +1,232 @@
+  }
+
+  auto schedule_decref = [](PyObject* obj) {
+    if (Py_AddPendingCall(+[](void* p){ Py_DECREF((PyObject*)p); return 0; }, obj) != 0) {
```
While `.has_value()` guarantees that a pointer is available, it doesn't guarantee its validity. Therefore, `if (!obj) return;` can be added before decrementing the reference.

sseshasa commented on src/test/mgr/test_mgrmapcache.cc:
```diff
@@ -0,0 +1,221 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:nil -*-
```
I think this commit can be squashed with the previous commit that introduces the MgMapCache. That way, the unit tests and the changes are in a single commit.

NitzanMordhai replied:
yes, it can be squashed

sseshasa reviewed (changes requested):
I added some more minor comments. But looks good now. Also there appears to be a related make check failure. PTAL.

sseshasa commented on src/mgr/MgrMapCache.cc:
```diff
@@ -0,0 +1,232 @@
+  ceph_assert(PyGILState_Check());
+  PyObject* o = nullptr;
+  if (!CacheImp::try_get(k, &o, true)) return nullptr;
+  Py_INCREF(o);                 // hand out a new ref
```
Looking closer, there could be a race here between the time `try_get()` returns (after releasing lock) and increment ref. This should be addressed I think.

NitzanMordhai replied:
right, i need to change that and not use try_get

sseshasa replied:
Looks good now.

NitzanMordhai commented:
@sseshasa fixed all the new points that you added! thanks a lot.
please check

sseshasa commented on src/mgr/MgrMapCache.cc:
```diff
@@ -0,0 +1,240 @@
+void MgrMapCache<PyObject*>::erase(std::string_view key) noexcept {
+  if (!this->is_cacheable(key)) return;
+  PyObject* o = nullptr;
+  if (!this->extract(key, &o)) return;
```
Is `o` guaranteed to be valid i.e. not a `nullptr` post `extract()`? If not, then an additional `nullptr` check before `Py_AddPendingCall()` would be safer.

NitzanMordhai replied:
No, as soon as extract() return with true, o have a valid value that should not be changed by Py_AddPendingCall.
No need to add an extra check for it

sseshasa reviewed (commented):
Possibly a last comment related to `erase()`. Can you PTAL? I will approve once this is addressed. Thanks!

sseshasa reviewed (approved):
LGTM. Just wanted to confirm if the test directories `mgr_ttl_cache` are also deleted from the tree? The files are deleted but am not sure about the directories. Otherwise, changes look good for qa! Thanks!

NitzanMordhai commented:
jenkins test make check

NitzanMordhai commented:
jenkins test make check arm64

NitzanMordhai commented:
/a/skanta-2026-05-26_00:49:28-rados-wip-bharath2-testing-2026-05-25-1222-distro-default-trial/237436
 
We are seeing error in testing only when we are called from dispatch_remote and cacheable_get_python is calling get cache
`PyObject* MgrMapCache<PyObject*>::get(std::string_view k) {
  if (!this->is_enabled() || !this->is_cacheable(k)) return nullptr;
  ceph_assert(PyGILState_Check());
  ` 
  The assertion will stop the processing instead of returning nullptr, and skipping the get cache, we should just skip the cache and get the map in that case.

NitzanMordhai commented:
Another issue that is related to the PR /a/skanta-2026-05-26_00:49:28-rados-wip-bharath2-testing-2026-05-25-1222-distro-default-trial/237446
We are missing a notify_all call when a new mgr_map is set, which causes the cache not to be updated. As a result, the upgrade test fails due to the absence of a standby manager after the upgrade.

batrick commented:
Please share regular updates on this `umbrella` branch blocker in `#ceph-upstream-releases-umbrella` on Slack.

yaarith commented on src/pybind/mgr/telemetry/module.py:
```diff
@@ -748,8 +748,7 @@ def get_io_rate(self) -> dict:
     def get_stats_per_pool(self) -> dict:
-        result = self.get('pg_dump')['pool_stats']
-
+        result = self.get('pool_stats', mutable=True)
```
let's make sure `mutable=True` does not change the original data

rzarzynski commented:
jenkins test api

phlogistonjohn commented:
Why does the first patch "randomly" change ceph.spec.in?

NitzanMordhai commented:
> Why does the first patch "randomly" change ceph.spec.in?
@phlogistonjohn That was a mistake. I removed it. I thought I could use it under teuthology, i'll leave it for later

NitzanMordhai commented:
jenkins test api

phlogistonjohn commented:
OK, thanks for the info.  For what it's worth Kefu is working on a refactor of the module packaging in #68989 and part of that includes. So it may be worth waiting for that PR if that's something you need packaged.

NitzanMordhai committed:
mgr/dashboard: cached osd_map pop pg_temp

get('osd_map') returns the cached object directly, so del and key
assignments were silently corrupting the cache for subsequent callers.
Take a shallow copy before modifying, and use pop() instead of del in
case the cache was already corrupted.

Fixes: https://tracker.ceph.com/issues/72447
Signed-off-by: Nitzan Mordechai <nmordec@ibm.com>

nizamial09 commented on src/pybind/mgr/dashboard/controllers/health.py:
```diff
@@ -296,10 +296,10 @@ def mon_status(self):
     def osd_map(self):
-        osd_map = mgr.get('osd_map')
+        osd_map = dict(mgr.get('osd_map'))
         assert osd_map is not None
```
this assert is obsolete now, but i don't think this will cause any functional regression, but just pointing out.
```suggestion
```

nizamial09 reviewed (commented):
lgtm from the @ceph/dashboard

NitzanMordhai commented:
> OK, thanks for the info. For what it's worth Kefu is working on a refactor of the module packaging in #68989 and part of that includes. So it may be worth waiting for that PR if that's something you need packaged.

Thanks for that info, but need that to Umbrella, i'll leave that tests outside for now (keep only the unit-test and workunit) and add them later on when it available

NitzanMordhai commented:
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrackercephcomissues76787
